Add REST API endpoint for shipping providers

Ferdev opened the pull request:
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding
Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR implements a REST API endpoint to expose shipping provider information for the CIAB Next Admin fulfillment workflow.

**Implementation Details:**
- Adds `/providers` route to the existing v4 Fulfillments Controller at `/wc/v4/fulfillments/providers`
- Returns shipping provider data including labels, icons, tracking URL templates
- Uses existing `FulfillmentUtils::get_shipping_providers_object()` method
- Requires `manage_woocommerce` capability for access

**Related Files:**
- `src/Internal/RestApi/Routes/V4/Fulfillments/Controller.php` - Added providers route and methods
- `src/Internal/Fulfillments/FulfillmentsController.php` - Registered ProvidersRestController (legacy, can be removed)
- `src/Internal/Fulfillments/ProvidersRestController.php` - Created standalone controller (unused, can be removed)
- `tests/php/src/Internal/RestApi/Routes/V4/Fulfillments/ProvidersTest.php` - Complete test coverage

Closes [WOOSHIP-1574](https://linear.app/a8c/issue/WOOSHIP-1574)

### Screenshots or screen recordings:

API Response Example:

    {
      "usps": {
        "label": "USPS",
        "icon": "https://example.com/.../usps.png",
        "value": "usps",
        "url": "https://tools.usps.com/go/TrackConfirmAction?tLabels=__PLACEHOLDER__"
      },
      "fedex": {
        "label": "FedEx",
        "icon": "https://example.com/.../fedex.png",
        "value": "fedex",
        "url": "https://www.fedex.com/fedextrack/?tracknumbers=__PLACEHOLDER__"
      }
    }

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions):

1. As an admin user, make a GET request to `/wp-json/wc/v4/fulfillments/providers`
2. Verify response status is 200
3. Verify response contains provider objects with `label`, `icon`, `value`, and `url` fields

### Testing that has already taken place:

### Changelog entry

- [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

- [ ] Patch
- [x] Minor
- [ ] Major

#### Type

- [ ] Fix - Fixes an existing bug
- [x] Add - Adds functionality
- [ ] Update - Update existing functionality
- [ ] Dev - Development related task
- [ ] Tweak - A minor adjustment to the codebase
- [ ] Performance - Address performance issues
- [ ] Enhancement - Improvement to existing functionality

#### Message

Add REST API endpoint `/wc/v4/fulfillments/providers` to expose shipping provider information including labels, icons, and tracking URLs for CIAB Next Admin integration.

</details>

Ferdev committed:
Add REST API endpoint for shipping providers

Implements /wc/v4/fulfillments/providers endpoint to expose shipping
provider information (labels, icons, tracking URLs) for the CIAB
Next Admin fulfillment workflow.

Changes:
- Add /providers route to v4 Fulfillments Controller
- Add get_providers() method returning FulfillmentUtils data
- Add permission checks requiring manage_woocommerce capability
- Register ProvidersRestController in FulfillmentsController

Ferdev committed:
Merge branch 'trunk' into wooship-1574-shipping-fulfillments-create-endpoint-to-get-shipping-providers

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

Ferdev committed:
Remove duplicated code

Ferdev committed:
Merge branch 'trunk' into wooship-1574-shipping-fulfillments-create-endpoint-to-get-shipping-providers

Ferdev committed:
Merge branch 'trunk' into wooship-1574-shipping-fulfillments-create-endpoint-to-get-shipping-providers

ayushpahwa commented on plugins/woocommerce/src/Internal/RestApi/Routes/V4/Fulfillments/Controller.php:
```diff
@@ -722,4 +735,25 @@ private function prepare_error_response( $code, $message, $data ): WP_REST_Respo
+	 */
+	public function get_providers( WP_REST_Request $request ): WP_REST_Response {
+		$providers = \Automattic\WooCommerce\Internal\Fulfillments\FulfillmentUtils::get_shipping_providers_object();
+		return new WP_REST_Response( $providers, WP_Http::OK );
```
I believe we should be using the `prepare_item_for_response` for formatting the response of our V4 APIs. It will handle the coherence to the schema and throw the response codes. The method is defined in the `AbstractController`.

ayushpahwa replied:
Refs
1) [Separated Schema Classes](https://wooengineering.wordpress.com/2025/09/22/developing-rest-api-v4-routes/#3-centralised-error-handling)
2) [Centralised Error Handling](https://wooengineering.wordpress.com/2025/09/22/developing-rest-api-v4-routes/#3-centralised-error-handling)

ayushpahwa commented on plugins/woocommerce/src/Internal/RestApi/Routes/V4/Fulfillments/Controller.php:
```diff
@@ -722,4 +735,25 @@ private function prepare_error_response( $code, $message, $data ): WP_REST_Respo
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response
+	 */
+	public function get_providers( WP_REST_Request $request ): WP_REST_Response {
```
I don't think we are using this arg here. Maybe we can remove it?

ayushpahwa replied:
@Ferdev this doesn't seem to be addressed.

ayushpahwa commented on plugins/woocommerce/src/Internal/RestApi/Routes/V4/Fulfillments/Controller.php:
```diff
@@ -722,4 +735,25 @@ private function prepare_error_response( $code, $message, $data ): WP_REST_Respo
+	 * @return bool
+	 */
+	public function check_permission_for_providers( WP_REST_Request $request ): bool {
+		return current_user_can( 'manage_woocommerce' );
```
This is expected as it is an admin only feature

ayushpahwa commented on plugins/woocommerce/tests/php/src/Internal/RestApi/Routes/V4/Fulfillments/ProvidersTest.php:
```diff
@@ -0,0 +1,255 @@
+		$response = rest_get_server()->dispatch( $request );
+		$data     = $response->get_data();
+
+		$this->assertIsArray( $data );
```
Would it make sense to add more validation here like the object structure etc.

ayushpahwa replied:
Oh I see another test for that, seems like this would be redundant in that case?

ayushpahwa commented on plugins/woocommerce/src/Internal/RestApi/Routes/V4/Fulfillments/Controller.php:
```diff
@@ -722,4 +735,25 @@ private function prepare_error_response( $code, $message, $data ): WP_REST_Respo
+	 * @return bool
+	 */
+	public function check_permission_for_providers( WP_REST_Request $request ): bool {
+		return current_user_can( 'manage_woocommerce' );
```
This check can return a better standardised error using the abstract class function 

`return $this->get_authentication_error_by_method( $request->get_method() );`

If not, we should remove the arg since it is not used here.

ayushpahwa replied:
This would be the sample output in this case

```JSON
{
  "code": "woocommerce_rest_api_v4_fulfillments_cannot_view",
  "message": "Sorry, you cannot view resources.",
  "data": {
    "status": 401
  }
}
```

ayushpahwa reviewed (commented):
Testing instructions work as expected and the tests look great! I had some V4 consistency related comments. Thanks for working on this 💯

ayushpahwa commented:
Another thing I see that might be out of scope of this PR but very easy to fix is related to the dependency injection. In the constructor of the controller we are calling the constructor of a dependency which can be changed to follow the dependency injection method defined [here](https://wooengineering.wordpress.com/2025/09/22/developing-rest-api-v4-routes/#1-dependency-injection)

Ferdev committed:
Merge branch 'trunk' into wooship-1574-shipping-fulfillments-create-endpoint-to-get-shipping-providers

Ferdev committed:
Merge branch 'trunk' into wooship-1574-shipping-fulfillments-create-endpoint-to-get-shipping-providers

Ferdev committed:
Address PR comments

Ferdev committed:
Merge branch 'trunk' into wooship-1574-shipping-fulfillments-create-endpoint-to-get-shipping-providers

Ferdev committed:
Fix linting problems

Ferdev committed:
Merge branch 'trunk' into wooship-1574-shipping-fulfillments-create-endpoint-to-get-shipping-providers

Ferdev committed:
Merge branch 'trunk' into wooship-1574-shipping-fulfillments-create-endpoint-to-get-shipping-providers

Ferdev committed:
Merge branch 'trunk' into wooship-1574-shipping-fulfillments-create-endpoint-to-get-shipping-providers

Ferdev committed:
Merge branch 'trunk' into wooship-1574-shipping-fulfillments-create-endpoint-to-get-shipping-providers

ayushpahwa commented on plugins/woocommerce/src/Internal/RestApi/Routes/V4/Fulfillments/Controller.php:
```diff
@@ -750,10 +752,58 @@ public function get_providers( WP_REST_Request $request ): WP_REST_Response {
+		return true;
+	}
+
+	/**
```
As per the V4 guidelines, it is recommended to keep the schema separately. The functions would be internally handled by the abstract controller. https://wooengineering.wordpress.com/2025/09/22/developing-rest-api-v4-routes/#2-separated-schema-classes

ayushpahwa reviewed (commented):
Thanks for making the changes, I think a comment got missed. Can you confirm? Thanks!

Ferdev committed:
Merge branch 'trunk' into wooship-1574-shipping-fulfillments-create-endpoint-to-get-shipping-providers

Ferdev committed:
Remove unused argument in fulfillments controller `get_providers`

Ferdev committed:
Refactor Fulfillments Controller to use separate Schema class

Following V4 REST API guidelines, moved schema definitions to a dedicated
FulfillmentSchema class extending AbstractSchema. This improves code
reusability and maintainability by:

- Creating FulfillmentSchema with get_item_schema_properties() and
  get_item_response() methods
- Replacing constructor with init() method for dependency injection
- Updating route registration to use standard V4 patterns
- Removing all inline schema methods from Controller

Ferdev committed:
Merge branch 'trunk' into wooship-1574-shipping-fulfillments-create-endpoint-to-get-shipping-providers

Ferdev committed:
Merge branch 'trunk' into wooship-1574-shipping-fulfillments-create-endpoint-to-get-shipping-providers

Ferdev committed:
Update ProvidersTest to initialize controller with schema

Add FulfillmentSchema initialization in setUp() to match the refactored
Controller that now requires init() to be called before register_routes().

ayushpahwa reviewed (approved):
The changes look great! The schema changes make the new file much cleaner! Thanks for working on this! :shipit:

Ferdev committed:
Merge branch 'trunk' into wooship-1574-shipping-fulfillments-create-endpoint-to-get-shipping-providers

Ferdev committed:
Add WP_REST_Request parameter to get_providers callback

The REST API passes request objects to callbacks, so the method
signature must accept it to avoid ArgumentCountError.

Ferdev committed:
Merge branch 'trunk' into wooship-1574-shipping-fulfillments-create-endpoint-to-get-shipping-providers

Ferdev committed:
Add response filter for shipping providers endpoint

Add woocommerce_rest_prepare_fulfillments_providers filter to allow
extensions to customize the providers response, following WooCommerce
REST API conventions.

Ferdev committed:
Inject OrderFulfillmentsRestController via DI pattern

Follow V4 controller convention by injecting dependencies through
init() method rather than direct instantiation.

Ferdev commented on plugins/woocommerce/src/Internal/RestApi/Routes/V4/Fulfillments/Controller.php:
```diff
@@ -49,12 +57,16 @@ class Controller extends AbstractController {
 	 */
-	public function __construct() {
-		$this->order_fulfillments_controller = new OrderFulfillmentsRestController();
+	final public function init( FulfillmentSchema $item_schema, OrderFulfillmentsRestController $order_fulfillments_controller ) {
```
This is not a backwards compatibility issue.

The class is in the `Automattic\WooCommerce\Internal` namespace, which is explicitly **not part of the public API**. Internal classes:

1. Are not subject to BC guarantees
2. Should not be extended or instantiated by third-party code
3. Are marked with `@internal` annotation (line 65)

The `init()` DI pattern is the standard approach for all V4 REST controllers in WooCommerce (see `Orders/Controller.php`, `Customers/Controller.php`, etc.). Extensions should use the public REST API endpoints, not instantiate internal controllers directly.

Ferdev commented on plugins/woocommerce/src/Internal/RestApi/Routes/V4/Fulfillments/Schema/FulfillmentSchema.php:
```diff
@@ -0,0 +1,150 @@
+						),
+						'value' => array(
+							'description' => __( 'The value of the meta data.', 'woocommerce' ),
+							'type'        => array( 'string', 'number', 'boolean', 'object', 'array', 'null' ),
```
This is a schema **correction**, not a breaking change.

The previous schema (`'type' => 'string'`) was **incorrect** - WordPress meta values can be any type (strings, arrays, objects, numbers, booleans, null). The new schema accurately
reflects what meta values can actually contain.

Additionally:
1. This class is in the `Internal` namespace - not public API
2. The Fulfillments feature is behind `woocommerce_feature_fulfillments_enabled` feature flag
3. The V4 Fulfillments API is not yet a stable/public API

Restricting meta values to strings only would break legitimate use cases where plugins store arrays or objects as meta data.

Ferdev committed:
Fix visibility of get_schema_for_providers REST callback

Change from private to public to allow REST server to invoke
the schema callback for OPTIONS requests and API discovery.

Ferdev committed:
Merge branch 'trunk' into wooship-1574-shipping-fulfillments-create-endpoint-to-get-shipping-providers

Ferdev committed:
Add validation for shipping providers filter response

- Document required provider structure in filter docblock
- Validate filtered result is an array
- Remove providers missing required keys (label, icon, value, url)
- Log _doing_it_wrong() when extensions return invalid structures

This helps maintain API stability by ensuring the response always
matches the documented schema regardless of filter modifications.

Ferdev committed:
Fix linting errors in providers validation

- Use esc_html__() for _doing_it_wrong() messages
- Fix variable alignment in validate_providers_structure()

Ferdev commented on plugins/woocommerce/src/Internal/RestApi/Routes/V4/Fulfillments/Controller.php:
```diff
@@ -90,29 +107,54 @@ public function register_routes() {
 			),
 		);
+
+		// Register the route for getting shipping providers.
+		register_rest_route(
+			$this->namespace,
+			$this->rest_base . '/providers',
+			array(
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_providers' ),
+					'permission_callback' => array( $this, 'check_permission_for_providers' ),
+					'schema'              => array( $this, 'get_schema_for_providers' ),
+				),
+			)
+		);
```
This has been fixed

Ferdev commented on plugins/woocommerce/src/Internal/RestApi/Routes/V4/Fulfillments/Controller.php:
```diff
@@ -366,360 +404,108 @@ protected function get_schema(): array {
+		 *
+		 * @since 10.5.0
+		 */
+		$providers = apply_filters( 'woocommerce_rest_prepare_fulfillments_providers', $providers, $request );
```
This has been addressed

Ferdev committed:
Merge branch 'trunk' into wooship-1574-shipping-fulfillments-create-endpoint-to-get-shipping-providers
